src/main/scala/mesosphere/marathon/api/v2/LabelSelector.scala

```diff
@@ -0,0 +1,84 @@
+package mesosphere.marathon.api.v2
+
+import mesosphere.marathon.state.AppDefinition
+import org.apache.log4j.Logger
+
+import scala.util.control.NonFatal
+import scala.util.parsing.combinator.RegexParsers
+
+case class LabelSelector(key: String, fn: String => Boolean, value: List[String]) {
+  def matches(app: AppDefinition): Boolean = app.labels.contains(key) && fn(app.labels(key))
+
+}
+case class LabelSelectors(selectors: Seq[LabelSelector]) {
+  def matches(app: AppDefinition): Boolean = selectors.forall(_.matches(app))
+}
+
+/**
+  * Parse a label selector query.
+  * A label selector query has this format:
+  * query: selector {, selector}
+  * selector: existenceSelector | setSelector | equalsSelector
+  * existenceSelector: label
+  * equalsSelector: term ==|!= term
+  * setSelector: term in|notin set
+  * set: ( term {, term} )
+  * term: character sequence with character groups A-Za-z0-9. Any other character needs to be escaped with backslash.
+  *
+  * Examples:
+  * test == foo
+  * test != foo
+  * environment in (production, qa)
+  * tier notin (frontend, backend)
+  * partition
+  * \!\!\! in (\-\-\-, \&\&\&, \+\+\+)
+  * a in (t1, t2, t3), b notin (t1), c, d
+  *
+  */
+class LabelSelectorParsers extends RegexParsers {
+
+  private[this] val log = Logger.getLogger(getClass.getName)
+
+  //Allowed characters are A-Za-z0-9. All other characters can be used, but need to be escaped.
+  def term: Parser[String] = """(\Q\\E.|[A-Za-z0-9])+""".r ^^ { _.replaceAll("""\Q\\E(.)""", "$1") }
```

Comment:
For me, `\Q\\E` is a pretty fancy way to quote a `\`. Why not use `\\`?

It would probably make sense to pre-compile the pattern used in the `replaceAll`.

But otherwise pretty dense implementation of escaping. I like it :)